### PR TITLE
Extend swagger-models to support RefPath and RefResponse per Swagger Reuse Guidelines

### PR DESCRIPTION
**Pull request by Helmsdown**
The [reuse guideline](https://github.com/swagger-api/swagger-spec/blob/master/guidelines/REUSE.md) mentions two types of references which are not currently supported by swagger-core. 

[This](https://github.com/swagger-api/swagger-spec/blob/master/guidelines/REUSE.md#operations) section makes note of referencing operations. 

The example(s) given for "operation refs" are:
```json
{
  "/health": {
    "$ref": "http://localhost:8000/operations.json#/health"
  }
}
```

```json
{
  "health": {
    "get": {
      "tags": [
        "admin"
      ],
      "summary": "Returns server health information",
      "operationId": "getHealth",
      "produces": [
        "application/json"
      ],
      "parameters": [],
      "responses": {
        "200": {
          "description": "Health information from the server",
          "schema": {
            "$ref": "http://localhost:8000/models.json#/Health"
          }
        }
      }
    }
  }
}
```
When I read this section I think *"shouldn't we call these path references?"*. The first json file example has a REST path of "/health" as the key.

If anything we shouldn't we update the first example to be the following?

```json
{
  "swagger": "2.0"
  "paths": {
    "/health": {
      "$ref": "http://localhost:8000/operations.json#/health"
    }
  }
}
```
To this end, I went ahead and implemented PathRefs (and not OperationRefs), assuming I am correctly interpreting the spirit of the Reuse guide.

Additionally, [this](https://github.com/swagger-api/swagger-spec/blob/master/guidelines/REUSE.md#responses) section of the reuse guideline notes that response definitions should be reusable through references.

To this end I modified the swagger-models classes in question to fit this proposed [design](http://www.gliffy.com/go/publish/8487689). I tried to follow the established example of Model and RefModel.

This change will have non trivial impact on other swagger libraries (and anyone that has written custom code that interacts with the Path and Response objects). 

Specifically the following code will no longer compile:

```java
Path p = new Path();
Response r = new Response();
```
and will have to be changed to:
```java
Path p = new PathImpl();
Response r = new ResponseImpl();
```
 Any code, swagger or otherwise, that processes Path and Response objects will need to be aware there might be RefPath and RefResponse instances lurking about. Having written custom code that iterates over Swagger objects, I am sensitive to the impact these kinds of changes can have. However, I think the long term benefits of being able to compose specs from previously defined Paths and Response (in additional to the already supported ref types) outweighs any short term pain.

Feedback on my code is appreciated.

### CODE REVIEW

**Comment by webron** [comment]
Since we're in the process of migrating to TestNG (and migrating all the tests), can you maybe convert the *new* tests to TestNG? It'd just help us with less migration.

**Comment by Helmsdown** [comment]
Absolutely. Shouldn't take me too long.

**Comment by webron** [comment]
Much appreciated. And from your note, once this is merged it resolved #1022 as well, right?

**Comment by Helmsdown** [comment]
Yes indeed.

**Comment by Helmsdown** [comment]
@webron The tests have been updated to Java/TestNG. Let me know if there are additional changes you'd like.

**Comment by webron** [comment]
Thanks for the work, @Helmsdown. I'm going to leave the review to @fehguy on this one though.

**Comment by Helmsdown** [comment]
Sounds good.

**Comment by fehguy** [comment]
As discussed, we should avoid changing the pattern for models until a 2.0 release, since it'll break every usage of the `SwaggerModel` artifact.  While this is much better, I suggest we work around the interfaces and push the renaming into next major.

**Comment by Helmsdown** [comment]
@fehguy Did as we we discussed. 

1. RefPath extends Path (until we can do a major release and make Path an interface)
2. RefResponse extends Response (*)
3. Added vendor extensions to the following swagger-models classes
 * AbstractModel.java
 * AbstractProperty.java
 * Response.java
 * SecurityRequirement.java
 * ExternalDocs.java
 * SecuritySchemeDefinition.java

Please note I had to do some "tomfoolery" in Json.java to support correct deserialization of RefPath/Path and RefResponse/Response. If you know of a better way to achieve the same result I would be very interested. Also, while I was messing around in Json.java I decided to do some light refactoring to central ize code that was duplicated between Json.java and Yaml.java.

As always, feedback on my code is greatly appreciated.

File: modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java [hunk 1]
```diff
@@ -1,11 +1,7 @@
 package io.swagger.core.filter;
 
 import io.swagger.model.ApiDescription;
-import io.swagger.models.Model;
-import io.swagger.models.Operation;
-import io.swagger.models.Path;
-import io.swagger.models.Swagger;
-import io.swagger.models.Tag;
+import io.swagger.models.*;
```
Comment (fehguy) [inline modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java]: In general, we've moved away from wildcard imports.  Not a blocker to this PR of course, just making sure you've seen the [style guides](https://github.com/swagger-api/swagger-core/wiki/Coding-Standards)

**Comment by fehguy** [comment]
Well this looks really good.  Let's get @webron's input as well as his spec changes first and we can see if this can be merged for the next minor release.

**Comment by Helmsdown** [comment]
@fehguy Auto-organize imports strikes again! Just reverted that particular change.

File: modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java (hunk 1 above)
Comment (Helmsdown) [inline modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java]: Auto-organize imports strikes again! Just reverted that particular change.

**Comment by webron** [comment]
After updating the spec, working out the proper places for vendor extensions, here is the full list:

- Swagger (i.e. the root object)
- Info
- Contact
- License
- Paths
- Path Item
- Operation
- External Docs
- Parameter
- Items
- Responses
- Response
- Header
- Tag
- Schema (which includes Property)
- XML
- Security Scheme
- Scopes

All other objects cannot contain vendor extensions. 

I suspect there may be an issue with the security-related objects in the Java implementation. Not sure if it's related to the change here or not, but perhaps should be taken into account. Have a look at #1258 and #1261.

**Comment by Helmsdown** [comment]
@webron Want me to refactor to satisfy #1258 and #1261? Are breaking changes ok? (since it apparently doesnt match the spec)

**Comment by webron** [comment]
Since it's breaking, I'd rather have @fehguy's opinion. It's a bit tricky.

**Comment by Helmsdown** [comment]
@fehguy Let me know what, if anything, you want me to do anything additional regarding #1258 and #1261...I have the time.

**Comment by webron** [comment]
Thanks for all the effort, @Helmsdown

**Comment by fehguy** [comment]
We can't break the models.  We can add to them though.  We can add new accessors or fields, but not break old ones :(

**Comment by webron** [comment]
How would we resolve #1258 and #1261 then?

**Comment by fehguy** [comment]
if we are breaking the old models then we cannot.  If we add new ones, then great.

**Comment by Helmsdown** [comment]
So #1261 sounds like it just needs a logic change. We need to add to this list of scopes...not override it with each successive call to *addSecurity()*...right?

#1258 sounds like it is a break change and we will have to wait till the next major version.

#1261 doesnt *need* to go in this pull request, but I can fix it if you like.

**Comment by Helmsdown** [comment]
You know on second read...I am confused by #1261...what exactly is the problem.

**Comment by webron** [comment]
#1261 shouldn't be part of this PR but requires #1258 to be solved. 
Not resolving #1258 means that there can't be proper security representation for tools that depend on models, and swagger-core can't be used in general to output such definitions.

**Comment by Helmsdown** [comment]
After re-reading #1258 again and looking at SecurityRequirement.java, I understand it now. 

This is what is allowed right now in Java
```java
        Swagger swagger = new Swagger();
        swagger.addSecurityDefinition(
                new SecurityRequirement("oauth")
                        .scope("scope1")
                        .scope("scope2"));
        swagger.addSecurityDefinition(
                new SecurityRequirement("api_key"));
```
and it gets serialized as
```json
{
  "swagger" : "2.0",
  "securityRequirement" : [ {
    "scopes" : [ "scope1", "scope2" ]
  }, { } ]
}
```
There are several issues at play:
1. The spec says this element should be called "security" not "securityRequirement" (can't fix this without a breaking chance)
2. The name of the requirement is not serialized due to a JsonIgnore annotation in SecurityRequirement.java
3. Even if the name of the requirement was serialized it wouldn't match the spec due to the the fact that the spec documents the format to be {"api_key": []} not {"name": "api_key", "scopes": ["scope1", "scope2"]
4. The java function to add a securityRequirement to the root swagger object is called *addSecurityDefintion()*

I have some proposed changes to remedy most of these with non breaking changes (and hopefully get us set up for cleaning this stuff up in the next major version). Will post the diff in a moment.

File: modules/swagger-models/src/main/java/io/swagger/models/SecurityRequirement.java [hunk 2]
```diff
@@ -1,19 +1,26 @@
...
 
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 public class SecurityRequirement {
     private String name;
     private List<String> scopes;
+//    private final Map<String, Object> vendorExtensions = new HashMap<String, Object>();
```
Comment (Helmsdown) [inline modules/swagger-models/src/main/java/io/swagger/models/SecurityRequirement.java]: can't have two @JsonAnyGetter/@JsonAnySetter maps. Being compliant with the primary purpose of this object seems more important than having vendor extensions in this object at this point.

File: modules/swagger-models/src/main/java/io/swagger/models/SecurityRequirement.java [hunk 3]
```diff
@@ -21,19 +28,40 @@ public SecurityRequirement scope(String scope) {
         return this;
     }
 
+    public SecurityRequirement requirement(String name, List<String> scopes) {
```
Comment (Helmsdown) [inline modules/swagger-models/src/main/java/io/swagger/models/SecurityRequirement.java]: support multiple definition references in this requirement

Comment (Helmsdown) [inline modules/swagger-models/src/main/java/io/swagger/models/SecurityRequirement.java]: support multiple references to security definitions in this requirement

File: modules/swagger-models/src/main/java/io/swagger/models/Swagger.java [hunk 4]
```diff
@@ -19,7 +20,7 @@
     protected List<Scheme> schemes;
     protected List<String> consumes;
     protected List<String> produces;
-    protected List<SecurityRequirement> securityRequirements;
+    protected List<SecurityRequirement> security;
```
Comment (Helmsdown) [inline modules/swagger-models/src/main/java/io/swagger/models/Swagger.java]: rename the private variable to bring in line with the spec definition

File: modules/swagger-models/src/main/java/io/swagger/models/Swagger.java [hunk 5]
```diff
@@ -115,6 +116,11 @@ public Swagger model(String name, Model model) {
         return this;
     }
 
+    public Swagger security(SecurityRequirement securityRequirement) {
```
Comment (Helmsdown) [inline modules/swagger-models/src/main/java/io/swagger/models/Swagger.java]: helper requested by #1258

File: modules/swagger-models/src/main/java/io/swagger/models/Swagger.java [hunk 6]
```diff
@@ -267,19 +273,39 @@ public void addSecurityDefinition(String name, SecuritySchemeDefinition security
...
     public void addSecurityDefinition(SecurityRequirement securityRequirement) {
-        if (this.securityRequirements == null) {
-            this.securityRequirements = new ArrayList<SecurityRequirement>();
+        if (this.security == null) {
+            this.security = new ArrayList<SecurityRequirement>();
+        }
+        this.security.add(securityRequirement);
+    }
+
+    @JsonIgnore //remove JsonIgnore when getSecurityRequirement() method is deleted in next major version
```
Comment (Helmsdown) [inline modules/swagger-models/src/main/java/io/swagger/models/Swagger.java]: don't want to serialize the security requirements twice

**Comment by Helmsdown** [comment]
Ok I think I have things, with respect to #1258, as good as they are going to get without a breaking change. Its not completely compliant with the spec (i.e. The list of security requirements is going to get serialized/deserialized as "securityRequirements" but the spec says it should be "security"). 

I @Deprecated all the methods that are wrong, but we can't delete because it would be breaking. 

We now should be able to model what is described in the spec. However, I *did not* change any of the code in swagger-core that actually acts on the @Authorization annotations (e.g. Reader.java).

@fehguy & @webron take a look at let me know if you want anything additional.

**Comment by fehguy** [comment]
looks good @Helmsdown

**Comment by Helmsdown** [comment]
Danke!

**Comment by FossiFoo** [comment]
Nice. Not breaking the old behaviour is fine with me, as long as we can implement the correct usage in generators with both "security" and "securityRequirements" for now. Then the "securityRequirements" one can be deprecated later in the next major.

**Pull request merged**